numpy/_core/src/multiarray/dlpack.c

```diff
@@ -32,16 +38,40 @@ array_dlpack_deleter(DLManagedTensor *self)
     PyGILState_Release(state);
 }
 
-/* This is exactly as mandated by dlpack */
-static void dlpack_capsule_deleter(PyObject *self) {
-    if (PyCapsule_IsValid(self, NPY_DLPACK_USED_CAPSULE_NAME)) {
+/* TODO: Basically same as above until dlpack v0 is removed: */
+static void
+array_dlpack_deleter_unversioned(DLManagedTensor *self)
+{
+    if (!Py_IsInitialized()) {
         return;
     }
 
-    DLManagedTensor *managed =
-        (DLManagedTensor *)PyCapsule_GetPointer(self, NPY_DLPACK_CAPSULE_NAME);
+    PyGILState_STATE state = PyGILState_Ensure();
+
+    PyArrayObject *array = (PyArrayObject *)self->manager_ctx;
+    PyMem_Free(self);
+    Py_XDECREF(array);
+
+    PyGILState_Release(state);
+}
+
+
+/*
+ * Deleter for a DLPack capsule wrapping a DLManagedTensor(Versioed).
+ *
+ * This is exactly as mandated by dlpack
+ */
+static void
+dlpack_capsule_deleter(PyObject *self) {
+    if (PyCapsule_IsValid(self, NPY_DLPACK_VERSIONED_USED_CAPSULE_NAME)) {
+        return;
+    }
+
+    DLManagedTensorVersioned *managed =
+        (DLManagedTensorVersioned *)PyCapsule_GetPointer(
+            self, NPY_DLPACK_VERSIONED_CAPSULE_NAME);
     if (managed == NULL) {
-        PyErr_WriteUnraisable(self);
```

Comment:
Doesn't this need the GIL?

Reply:
No, a capsule is a Python object, so it can't be deleted without the GIL.  It is just the DLManagedTensor deleter that needs to grab the GIL.